rfc/instance.md

```diff
@@ -0,0 +1,145 @@
+
+### The polling optimization
+
+The current `IntersectionObserver` polyfill checks every observed element every frame to see if the intersection state of an element has changed. Using requestAninmationFrame to poll on every frame requires significant CPU utilization. Meanwhile, it's not actually necessary to poll on every frame. If there were no changes to the window (scroll/resize) and no DOM changes, then there is no need to re-calculate intersection state.
+
+The solution should:
+- Fire all the events as expected/ as it was before.
+- Reduce CPU and memory utilization.
+- Should not create extra overhead while the system is idle.
+
+# Detailed design
+Rather than polling the viewport status of every time on every frame using `requestAnimationFrame`, Spaniel should just poll on the next animation frame after a scroll, resize or DOM change event occurs. To achieve this, we need to listen for scroll, resize and DOM change events, with respect to the viewport/[root](https://wicg.github.io/IntersectionObserver/#intersectionobserver-intersection-root). 
+
+We can listen for scroll or resize events using `addEventListener` function. But there is no universal way to listen for DOM changes, as [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) isn't universal. So we must provide an API through which the host app can provide a hook that allows Spaniel to listen for DOM changes.
+
+## Proposed API
+
+```javascript
+import { SpanielInstance } from 'spaniel';
+const spanielInstance = new SpanielInstance({
+  IntersectionObserver,
+  engine : {
+     onMutate,
+     schedulework
+  }
+});
+```
+- **IntersectionObserver**: This will allow user to spacify what IntersectionObserver user wants to use in this spaniel instance. User can have his own implementation of IntersectionObserver or pass the native intersectionObserver. If this argurment is skipped, we will use native Intersection observer if the browser supports or we will use spaniel implementation of IntersectionObserver.
+
+- **onMutate** : A function which fires when DOM modification occurrs.
+- **schedulework**: A function to populate work queue which is to be processed in rFA callback. If we have this argument constructor will replace spaniel implementation of the schedulework. Few frameworks need all the work to be executed inside their own system. Then we need this argument.
+
+
+Only when user provides `engine` object with `onMutate` function we create an `Optimized Engine`, which will use scroll/resize/DOM change events to process scheduled work and read queues in the next earliest rFA cycle. Otherwise constructor will fallback to creating normal implementation of engine which will poll rFA very frequently.
```

Comment:
This paragraph is sort hard to read.  Would the following be accurate?

If an `onMutate` callback is provided, the given engine instance will schedule work only after outside signals (such as scroll/resize/DOM change) are observed. This aims to mitigate the cost of constant rAF polling. Alternatively, if no `onMutate` callback is provided, the given engine instance will fallback to rAF based polling.